### PR TITLE
ghidra: add findcrypt extension

### PR DESCRIPTION
**Pull request by BonusPlay**
## Description of changes

Ghidra FindCrypt is a plugin to aid in locating cryptographic constants within analyzed binaries.
There are multiple plugins with this functionality, so I've chosen the most recent and up-to-date one.
I have been using this extension for the past ~2 months, so I've decided to upstream it to nixpkgs.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/tools/security/ghidra/extensions/findcrypt/default.nix [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    owner = "antoniovazquezblanco";
+    repo = "GhidraFindcrypt";
+    rev = "v${version}";
+    hash = "sha256-FSRuZtVEErEo4z134XRSsYcIccs/F41R2ydnBy+992o=";
+  };
+
+  meta = {
+    description = "Ghidra analysis plugin to locate cryptographic constants";
+    homepage = "https://github.com/antoniovazquezblanco/GhidraFindcrypt";
+    downloadPage = "https://github.com/antoniovazquezblanco/GhidraFindcrypt/releases/tag/v3.0.1";
```
Comment (h7x4) [inline pkgs/tools/security/ghidra/extensions/findcrypt/default.nix:22]:
```suggestion
    downloadPage = "https://github.com/antoniovazquezblanco/GhidraFindcrypt/releases/tag/v${version}";
```

File: pkgs/tools/security/ghidra/extensions/findcrypt/default.nix [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  src = fetchFromGitHub {
+    owner = "antoniovazquezblanco";
+    repo = "GhidraFindcrypt";
+    rev = "v${version}";
+    hash = "sha256-FSRuZtVEErEo4z134XRSsYcIccs/F41R2ydnBy+992o=";
+  };
+
+  meta = {
+    description = "Ghidra analysis plugin to locate cryptographic constants";
+    homepage = "https://github.com/antoniovazquezblanco/GhidraFindcrypt";
```
Comment (h7x4) [inline pkgs/tools/security/ghidra/extensions/findcrypt/default.nix:21]:
```suggestion
    homepage = "https://github.com/antoniovazquezblanco/GhidraFindcrypt";
    changelog = "https://github.com/antoniovazquezblanco/GhidraFindcrypt/releases/tag/v${version}"
```

**Review by h7x4 — COMMENTED** [review]
Neat :+1:

File: maintainers/maintainer-list.nix [hunk 3]
```diff
@@ -23638,5 +23638,13 @@
...
+  BonusPlay = {
+    name = "Bonus";
+    email = "nixos@bonusplay.pl";
+    matrix = "@bonus:bonusplay.pl";
+    github = "BonusPlay";
+    githubId = 8405359;
+    keys = [ { fingerprint = "8279 6487 A4CA 2A28 E8B3  3CD6 C7F9 9743 6A20 4683"; } ];
+  };
 }
 # Keep the list alphabetically sorted.
```
Comment (h7x4) [inline maintainers/maintainer-list.nix:23650]: ^ you should probably move this a bit further up

Comment (BonusPlay) [reply maintainers/maintainer-list.nix]: yep, working on it :+1:

File: pkgs/tools/security/ghidra/extensions/findcrypt/default.nix (hunk 2 above)
Comment (BonusPlay) [reply pkgs/tools/security/ghidra/extensions/findcrypt/default.nix:22]: Added :+1:

File: pkgs/tools/security/ghidra/extensions/findcrypt/default.nix (hunk 1 above)
Comment (BonusPlay) [reply pkgs/tools/security/ghidra/extensions/findcrypt/default.nix]: Added :+1:

**Comment by BonusPlay** [comment]
Sorry for the mess, needed to dust-off my git-foo.

File: pkgs/tools/security/ghidra/extensions.nix [hunk 4]
```diff
@@ -27,4 +27,6 @@ lib.makeScope newScope (self: {
 
   sleighdevtools = self.callPackage ./extensions/sleighdevtools { inherit ghidra; };
 
+  findcrypt = self.callPackage ./extensions/findcrypt { };
```
Comment (vringar) [inline pkgs/tools/security/ghidra/extensions.nix:30]: We're keeping the extension sorted alphabetically so please move yours up

File: pkgs/tools/security/ghidra/extensions/findcrypt/default.nix [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  version = "3.0.1";
+in
+buildGhidraExtension {
+  pname = "findcrypt";
+  inherit version;
+
+  src = fetchFromGitHub {
+    owner = "antoniovazquezblanco";
+    repo = "GhidraFindcrypt";
+    rev = "v${version}";
```
Comment (vringar) [inline pkgs/tools/security/ghidra/extensions/findcrypt/default.nix:16]: Iirc I was told that the Rev should be a string literal. I think this is related to the auto update bot that generates PRs when a new version gets released.

**Review by vringar — CHANGES_REQUESTED** [review]
Hey,
Thanks for upstreaming this! 
Two quick things that should be easy to fix

File: pkgs/tools/security/ghidra/extensions/findcrypt/default.nix (hunk 5 above)
Comment (h7x4) [reply pkgs/tools/security/ghidra/extensions/findcrypt/default.nix:16]: I don't think I've ever seen this recommendation before. I know rev is usually set as a string literal when you're working with unstable versions, but I'd love to know more about this, and maybe even add it to https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#sources. Do you happen to remember where you got it from?

Comment (BonusPlay) [reply pkgs/tools/security/ghidra/extensions/findcrypt/default.nix:16]:
```
$ grep -Rni rev | grep -Rni '${version}' | wc -l
25517
```
I'm pretty sure this pattern is pretty common...

Comment (vringar) [reply pkgs/tools/security/ghidra/extensions/findcrypt/default.nix:16]: Turns out I remembered it [the wrong way around](https://github.com/NixOS/nixpkgs/pull/315672#discussion_r1699211138) 
sorry for the useless comment, it was a drive-by review from my phone and I couldn't check.

**Comment by vringar** [comment]
When running `nixpkgs-review pr 343256` I get the following error:
```
error: hash mismatch in fixed-output derivation '/nix/store/wzbbrnzzp451b6yh6lgb3ayz2w72q3a7-source.drv':
         specified: sha256-FSRuZtVEErEo4z134XRSsYcIccs/F41R2ydnBy+992o=
            got:    sha256-/KA95NdoQFvR6XSGCHkX+ySKfftK84hJ8zjAvt0+O0o=
error: 1 dependencies of derivation '/nix/store/361dj78gzjgryg3hrd0llhhhvd1hx6jj-findcrypt-3.0.1.drv' failed to build
error: 1 dependencies of derivation '/nix/store/6ml1yhx46paibb6nvk27mircn1ca1f08-review-shell.drv' failed to build
```
Can you reproduce this?

**Review by vringar — CHANGES_REQUESTED** [review]
When running `nixpkgs-review pr 343256` I get the following error:
```
error: hash mismatch in fixed-output derivation '/nix/store/wzbbrnzzp451b6yh6lgb3ayz2w72q3a7-source.drv':
         specified: sha256-FSRuZtVEErEo4z134XRSsYcIccs/F41R2ydnBy+992o=
            got:    sha256-/KA95NdoQFvR6XSGCHkX+ySKfftK84hJ8zjAvt0+O0o=
error: 1 dependencies of derivation '/nix/store/361dj78gzjgryg3hrd0llhhhvd1hx6jj-findcrypt-3.0.1.drv' failed to build
error: 1 dependencies of derivation '/nix/store/6ml1yhx46paibb6nvk27mircn1ca1f08-review-shell.drv' failed to build
```
Can you reproduce this?

**Comment by BonusPlay** [comment]
No, not really.
```
$ nix-shell -p nixpkgs-review --run "nixpkgs-review pr 343256"
$ git -c fetch.prune=false fetch --no-tags --force https://github.com/NixOS/nixpkgs master:refs/nixpkgs-review/0 pull/343256/head:refs/nixpkgs-review/1
$ git worktree add /home/bonus/.cache/nixpkgs-review/pr-343256-1/nixpkgs 9a0f009815f30d0753f1e4dda4274bcb70f255ad
Preparing worktree (detached HEAD 9a0f009815f3)
Updating files: 100% (43351/43351), done.
HEAD is now at 9a0f009815f3 ibus-engines.mozc-ut: init at 2.30.5544.102 (#314248)
$ git merge --no-commit --no-ff 4bbf4b95431fde921db405ad7f7bd4a71b563a51
Auto-merging maintainers/maintainer-list.nix
Automatic merge went well; stopped before committing as requested
$ nix build --file /nix/store/kz30bnfvzwl476hwq5fgbyfalm2ynhi1-nixpkgs-review-2.10.4/lib/python3.11/site-packages/nixpkgs_review/nix/review-shell.nix --nix-path nixpkgs=/home/bonus/.cache/nixpkgs-review/pr-343256-1/nixpkgs nixpkgs-overlays=/tmp/tmp7iw3by83 --extra-experimental-features nix-command no-url-literals --no-link --keep-going --no-allow-import-from-derivation --option build-use-sandbox relaxed --argstr system x86_64-linux --argstr nixpkgs-path /home/bonus/.cache/nixpkgs-review/pr-343256-1/nixpkgs --argstr nixpkgs-config-path /tmp/tmptxd_b6ul.nix --argstr attrs-path /home/bonus/.cache/nixpkgs-review/pr-343256-1/attrs.nix

Link to currently reviewing PR:
https://github.com/NixOS/nixpkgs/pull/343256

1 package built:
ghidra-extensions.findcrypt

$ /nix/store/5q78686d7y90zlfb9jj03q8wgj36im4g-nix-2.18.5/bin/nix-shell --argstr system x86_64-linux --argstr nixpkgs-path /home/bonus/.cache/nixpkgs-review/pr-343256-1/nixpkgs --argstr nixpkgs-config-path /tmp/tmptxd_b6ul.nix --argstr attrs-path /home/bonus/.cache/nixpkgs-review/pr-343256-1/attrs.nix --nix-path nixpkgs=/home/bonus/.cache/nixpkgs-review/pr-343256-1/nixpkgs nixpkgs-overlays=/tmp/tmp7iw3by83 /nix/store/kz30bnfvzwl476hwq5fgbyfalm2ynhi1-nixpkgs-review-2.10.4/lib/python3.11/site-packages/nixpkgs_review/nix/review-shell.nix
```

**Comment by BonusPlay** [comment]
However, running
```
$ nix run nixpkgs#nix-prefetch-github -- --nix antoniovazquezblanco GhidraFindcrypt --rev v3.0.1
let
  pkgs = import <nixpkgs> {};
in
  pkgs.fetchFromGitHub {
    owner = "antoniovazquezblanco";
    repo = "GhidraFindcrypt";
    rev = "06731313669c706fccf35fa8b9f9da5e79588bf0";
    hash = "sha256-/KA95NdoQFvR6XSGCHkX+ySKfftK84hJ8zjAvt0+O0o=";
  }
  ```
  shows that something weird is going on...

**Comment by h7x4** [comment]
@BonusPlay Maybe you have an existing FOD with the previous hash? Try building without the hash once more and see what it reports as expected

**Comment by BonusPlay** [comment]
Ok, apparently something got lost somewhere in translation, the `sha256-FSRuZtVEErEo4z134XRSsYcIccs/F41R2ydnBy+992o=` hash is for the latest commit `4699bce02b1aafa85ee4a7ab96d5dcbf0d7cc7eb` from `main` branch. Honestly, I have no idea how this could have happened, as the derivation built correctly before:
```
$ nix-build -E "with import <nixpkgs> {}; callPackage ./ghidra-findcrypt.nix {}"
/nix/store/x57w4yvacdqwzmw7hji3jr29fprmjwy0-findcrypt-11.0.3
```
I've updated this PR with correct hash for the `v3.0.1` release.

**Review by vringar — APPROVED** [review]

**Comment by vringar** [comment]
Builds and works for me :)

**Review by h7x4 — APPROVED** [review]

**Pull request merged**